File: 0441-arranging-coins/0441-arranging-coins.cpp
class Solution {
public:
    int arrangeCoins(int n) {
        long long low = 0, high = n;

        while (low <= high) {
            long long mid = low + (high - low) / 2;
            long long sum = mid * (mid + 1) / 2; 

            if (sum == n) {
                return mid; 
            } else if (sum < n) {
                low = mid + 1; 
            } else {
                high = mid - 1; 
            }
        }

        return high; 
    }
};



        // int num = n; 
        // int count = 1;

        // if(n == 1) return 1;
        // while(true){
        //     num = num - count;
        //     if(num == 0){
        //         return count;
        //     }
        //     else if(num < 0){
        //         return --count;
        //     }
        //     else{
        //         count++;
        //     }

File: 0441-arranging-coins/0441-arranging-coins.java
class Solution {
    public int arrangeCoins(int n) {
        long l = 0, r = n;
        while(l <= r){
            long mid = l + (r - l) / 2;
            long count = mid * (mid + 1) / 2;
            if(count == n) return (int)mid;
            if(count < n){
                l = mid + 1;
            }
            else{
                r = mid - 1;
            }
        }
        return (int)r;
    }
}

        // int row = 0;
        // while(row + 1 <= n){
        //     row++;
        //     n -= row;
        // }
        // return row;

        
        // int row = 0, count = 0, total = 0;
        // while(total + row + 1 <= n){
        //     row += 1;
        //     total += row;
        //     count++;
        // }
        // return count;